### PR TITLE
Enable database items in Resonant Laboratory

### PR DESCRIPTION
**Pull request by ronichoudhury**
This is WIP because it doesn't quite work yet. I will need some help tracking down the problems and possibly redesigning some of the datatype handling stuff.

### CODE REVIEW

**Comment by alex-r-bigelow** [comment]
After poking around with this a little, it looks like the problem is stemming from only using datalib to infer the attributes in the dataset (if it chokes, for now I stupidly assume that we can't figure out what the attributes are, hence the error that the interface is showing... BTW, if you merge the ui-fixes branch onto this one, the error message should work properly).

I was able to get this to work with some of the datasets we've already been working with, e.g.:
```
mongoimport --db test --collection gapminder --drop --file gapminder.json --jsonArray
```
Then, of course, create an item, and then give it its connection information via `PUT /item/{id}/database`:
```
{"url":"localhost:27017","database":"test","collection":"gapminder","type":"mongo"}
```
(you probably know most of this... this is more documentation for my benefit :-) )

To make this work in general, I think we just need to infer the attributes of the response in a more intelligent way... Actually, we should probably retrieve the *schema* of the dataset at some point anyway so that we can build the filter interface...

**Comment by ronichoudhury** [comment]
The gapminder data seems to work. Let me know what happens when you try out points.json using the same procedure.

**Comment by alex-r-bigelow** [comment]
I just got it to work... It may have something to do with my dumb reliance on the item name having an "extension" that reflects its format (I'll file an issue). If you rename the item to be "points.json" it should work.

**Comment by ronichoudhury** [comment]
Well I did name mine `points.json`... but in my Girder view of that item, I see it lists `{}` as its `attributes` metadata entry. Does that have something to do with it?

**Comment by alex-r-bigelow** [comment]
That's very possible. You could try trashing its metadata and see if it helps

**Comment by ronichoudhury** [comment]
Welp. That worked beautifully. I wonder what went wrong on Friday...

**Comment by alex-r-bigelow** [comment]
A lot of these problems seem to be related to this thread:
#216 #203 #175 #169 #168 #162

I've generalized #204 to try and describe the higher-level issue

File: app/resonant-laboratory/models/Dataset.js [hunk 1]
```diff
@@ -89,20 +89,26 @@ let Dataset = MetadataItem.extend({
...
-      }).catch(() => {
-        this.rawCache = null;
-        return null;
+        dataType: 'json'
       });
+
+      return databaseQuery.then(resp => {
+        return JSON.stringify(resp.data);
+      }, () => {
+        console.log('hello');
```
Comment (jeffbaumes) [inline app/resonant-laboratory/models/Dataset.js]: Remove debug statement.

Comment (ronichoudhury) [inline app/resonant-laboratory/models/Dataset.js]: Oops! Removed.

**Comment by alex-r-bigelow** [comment]
LGTM—for now, I think it's fine to probe database status, and let the 400 error show up on the console. I'll probably add a more direct check whether an item is a database vs a file as part of my sprint tomorrow

**Comment by ronichoudhury** [comment]
In the future the hiccup about "database items" should theoretically go away. It's something I need to discuss with Zach but eventually it will all just be abstract Girder items - no special probing logic necessary.

**Comment by alex-r-bigelow** [comment]
Are you planning on making database/not-database status more of a first-class attribute of an item (is it already)? I was planning on adding a flag to the item's metadata, but there's no sense in being redundant...

**Comment by ronichoudhury** [comment]
No, we shouldn't make explicit whether an item is database-backed or not; it should be an abstract feature of an item. Currently the only way to autodetect whether an item has database backing is probing the endpoint and watching for an error. If putting a flag on the item makes this logic easier/cleaner, then we should do that. But in any case both approaches reflect a flaw in the current architecture, which can be solved in the long term by introducing the correct database-backing semantics to Girder items.

If this doesn't make sense let's talk IRL :smile_cat:

**Pull request merged**